modules/conversion-center/classes/render/render-base.php

```diff
@@ -166,78 +198,217 @@ protected function render_bio(): void {
+					$bio_about_heading_output = sprintf( '<%1$s %2$s>%3$s</%1$s>', Utils::validate_html_tag( $bio_about_heading_props_tag ), $this->widget->get_render_attribute_string( 'about-heading' ), esc_html( $bio_about_heading_value ) );
+					// Escaped above
+					Utils::print_unescaped_internal_string( $bio_about_heading_output );
+				} ?>
+				<?php if ( $has_bio_description ) {
+					$this->widget->add_render_attribute( 'description', 'class', 'e-link-in-bio__description' );
+					$bio_description_output = sprintf( '<p %1$s>%2$s</p>', $this->widget->get_render_attribute_string( 'description' ), esc_html( $bio_description_value ) );
+					// Escaped above
+					Utils::print_unescaped_internal_string( $bio_description_output );
+				} ?>
+			</div>
+			<?php
+		}
+	}
+
+	protected function get_shape_divider( $side = 'bottom' ) {
+		$settings = $this->settings;
+		$base_setting_key = "identity_section_style_cover_divider_$side";
+		$file_name = $settings[ $base_setting_key ];
+		if ( empty( $file_name ) ) {
+			return [];
+		}
+		$negative = ! empty( $settings[ $base_setting_key . '_negative' ] );
+		$shape_path = Shapes::get_shape_path( $file_name, $negative );
+		if ( ! is_file( $shape_path ) || ! is_readable( $shape_path ) ) {
+			return [];
+		}
 
-		$has_identity_image = ! empty( $identity_image_value ) && ( ! empty( $identity_image_value['url'] || ! empty( $identity_image_value['id'] ) ) );
+		return [
+			'negative' => $negative,
+			'svg' => Utils::file_get_contents( $shape_path ),
+		];
+	}
 
-		if ( ! $has_identity_image ) {
+	protected function print_shape_divider( $side = 'bottom' ) {
+		$shape_divider = $this->get_shape_divider( $side );
+		if ( empty( $shape_divider ) ) {
 			return;
 		}
+		?>
+		<div
+			class="elementor-shape elementor-shape-<?php echo esc_attr( $side ); ?>"
+			data-negative="<?php
+				echo esc_attr( $shape_divider['negative'] ? 'true' : 'false' );
+			?>"
+		>
+			<?php
+				// PHPCS - The file content is being read from a strict file path structure.
+				echo $shape_divider['svg']; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+			?>
+		</div>
+		<?php
+	}
 
-		$identity_image_classnames = 'e-link-in-bio__identity-image e-link-in-bio__identity-image-' . $identity_image_props_style;
-		if ( $identity_image_props_show_border || $identity_image_props_show_bottom_border ) {
-			$identity_image_classnames .= ' has-border';
+	protected function render_identity_image(): void {
+		/**
+		 * Get base data for potential images
+		 * Note order is important - secondary must render before primary
+		*/
+		$output_images = [
+			'secondary_image' => [
+				'props' => [],
+				'should_render' => false,
+				'value' => $this->settings['identity_image_cover'] ?? [],
+			],
+			'primary_image' => [
+				'props' => [],
+				'should_render' => false,
+				'value' => $this->settings['identity_image'] ?? [],
+			],
+		];
+
+		// Bail early if there are no images to render
```

Comment:
This looks a bit messy, and it difficult to read.

Maybe extract to a standalone function and possibly try to structure the code more?

Reply:
@hein-obox refactored